### PR TITLE
[Patch 1/2] Feature/client capability response [Test 4.6.2]

### PR DESCRIPTION
**Pull request by CoralMalachi**
### Description
According to Multi-AP Specification, If a Multi-AP Agent receives a
Client Capability Query message, then within one second it shall respond
with a Client Capability Report message.

If the STA specified in the Client Capability Query message is not
associated with any of the BSS operated by the Multi-AP Agent (an error
scenario), the Multi-AP Agent shall set the result code in the Client
Capability Report TLV to 0x01  and include an Error Code TLV.

### Changes
Therefore, the following changes were made:

- Create a Client Capability Report message CMDU.

- Add the Client Info TLV.

- Add a Client Capability Report TLV.

- Check that the Multi-AP Agent has the BSSID that was specified in the Client Info TLV, and that the client is associated with that BSS. If not, this is an error scenario, so set the result code in the Client
Capability Report TLV to 0x01, and include an Error Code TLV with the reason code set to 0x02.

### Testbed status
This branch was tested on testbed while using Qualcomm as CTT Controller, Marvell STA as client and prplMesh as MAUT.
Currently, there is an issue with the Marvell STA, which can not connect to the prplMesh agent.
Since the client was not connected to the agent, the Client Capability Report contained an Error Code TLV as it should.

_Status update: test 4.6.2 **passed** the test on the testbed._

### Next steps

- [x] After solving the issue with the Marvell STA, run the test again and verify that in case of success the Client Capability Report does not contain an Error Code TLV.

- [ ] Client Capability Report TLV  should include the frame body of the most recently received (Re)Association Request frame from this client (if it is an error scenario this field is omitted).
This part will be done in a separate PR.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

### CODE REVIEW

**Review by tomereli — APPROVED** [review]
Some comments I'd like your feedback on before merge BUT nothing serious so approving.

File: common/beerocks/bcl/include/bcl/network/network_utils.h
Comment (arnout) [inline common/beerocks/bcl/include/bcl/network/network_utils.h]: This is only needed because you're using a std::set or std::map somewhere, which you shouldn't do.

File: agent/src/beerocks/slave/son_slave_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: std::unordered_set

File: agent/src/beerocks/slave/son_slave_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: We might want to throw a warning if the vap is not found...

File: agent/src/beerocks/slave/son_slave_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: This is not accurate... If the client is associated with a BSS on another radio in the same agent, both radios will receive the query, but it is not associated with this radio. We shouldn't treat that as an error.

Also, the bssid given in the client info should be treated as just a hint. The spec says "If the STA specified in the Client Capability Query message is not associated with any of the BSS operated by the Multi-AP Agent". It doesn't say "if it is not associated with the BSS given by the client info TLV". A likely scenario is that a client associates and immediately disassociates and moves to another band, so it's possible that the controller sends the query still using the old BSS (which it got from the topology notification).

Bottom line, I think the best approach is to again cache everything on the backhaul_manager and generate the message there.

File: agent/src/beerocks/slave/son_slave_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: bssid may have changed already...

For all we know, stupid controllers may just leave it at 00:00:00:00:00:00.

File: controller/src/beerocks/master/son_master_thread.cpp
Comment (arnout) [inline controller/src/beerocks/master/son_master_thread.cpp]: Please also log the details so we can check it in test_flows.

**Review by arnout — CHANGES_REQUESTED** [review]
I unfortunately have some fundamental architectural comments:

> Bottom line, I think the best approach is to again cache everything on the backhaul_manager and generate the message there.

To be verified with @tomereli if he agrees.

File: agent/src/beerocks/slave/son_slave_thread.cpp
Comment (CoralMalachi) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: Thanks for pointing it, I'll change it

**Comment by tomereli** [comment]
> I unfortunately have some fundamental architectural comments:
> 
> > Bottom line, I think the best approach is to again cache everything on the backhaul_manager and generate the message there.
> 
> To be verified with @tomereli if he agrees.

I completely agree and I think I also suggested that in my review, didn't want to block merge on this but on second thought I think you are right, we should spend the extra effort and start using the VAPs map for everything - clients, client capabilities, etc.

File: controller/src/beerocks/master/son_master_thread.cpp
Comment (CoralMalachi) [reply controller/src/beerocks/master/son_master_thread.cpp]: Client Capability Report TLV should include the frame body of the most recently received (Re)Association Request frame from this client (if it is an error scenario this field is omitted).
This part will be done in a separate PR.
So, for now the log message can contain if this is an error scenario or not, and I'll add a TODO comment to add the details later.

File: controller/src/beerocks/master/son_master_thread.cpp
Comment (arnout) [inline controller/src/beerocks/master/son_master_thread.cpp:988]: :1st_place_medal:

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: It may be useful to explicitly mention that even though client_info_tlv_r does contain a BSSID, the specification says that we should answer if the client is associated with *any* BSS on this agent.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Use `true`/`false` for booleans, not 1 and 0.

Also, instead of `error_scenario`, I think it makes more sense to call it e.g. `associated_client_found` (and inverse the logic, obviously).

(nitpick)

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1871]: So, this...

We have two options here. Either we send a message to the slave that forwards it to the ap_manager that converts it to a bwl call and generates a response message to the slave that gets forwarded to the backhaul manager that generates the response message (with the same MID). Or, we add the ReAssociation Request frame to the CLIENT_ASSOCIATED_NOTIFICATION and cache it in the backhaul manager. The latter would also require us to resend the notification every time a new reassociation request is received by the AP.

If we choose the first option, we have to take into account a possible race condition: by the time the message reaches hostapd, the client may already have left the BSS.

@vitalybu @morantr @tomereli your opinions?

Note that for this PR it doesn't matter much. If it is cached, we'll just need to update the loop in the beginning that finds the associated client to not just set a flag, but to set (a pointer to) the association request frame.

File: tests/test_flows.sh
Comment (arnout) [inline tests/test_flows.sh]: Please use send_CAPI_1905 function.

File: tests/test_flows.sh
Comment (arnout) [inline tests/test_flows.sh]: It would be better to check this on the controller instead of the agent. The commit that adds the handler on the controller could very easily parse out the result code and report it.

Also, to make sure the correct log message is checked, add the client MAC and BSSID to the output.

File: tests/test_flows.sh
Comment (arnout) [inline tests/test_flows.sh]: Again, send_CAPI_1905 please.

File: tests/test_flows.sh
Comment (arnout) [inline tests/test_flows.sh]: Same here, better to check it on the controller.

File: tests/test_flows.sh
Comment (arnout) [inline tests/test_flows.sh]: Please use a *different* sta_mac here than the one for the success scenario. This helps to make sure we parse out the correct log message. Cfr. #822

File: tests/test_flows.sh
Comment (arnout) [inline tests/test_flows.sh]: You can actually use `${sta_mac}` here - our CAPI parser accepts it.

**Review by arnout — APPROVED** [review]
*Very* good splitting in commits and good commit messages. :1st_place_medal: 

Of course I have a few nitpicks, but we can go through with merging this.

There is one major question remaining about how it should be implemented.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (CoralMalachi) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Agree.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (tomereli) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1871]: I prefer option (2) - caching. I think it is cleaner.
However, the race condition may occur in both options since the client may have left also by the time the message with the updated ReAssociation frame is received in the BH manager

Comment (CoralMalachi) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1871]: I also prefer the second option.

**Pull request merged**
